app/server/lib/GristServerSocket.ts

```diff
@@ -0,0 +1,142 @@
+
+  public set onerror(handler: (err: unknown) => void) {
+    // Note that as far as I can tell, Engine.IO sockets never emit "error"
+    // but instead include error information in the "close" event.
+    this._socket.on('error', handler);
+    this._eventHandlers.push({event: 'error', handler});
+  }
+
+  public set onclose(handler: () => void) {
+    const wrappedHandler = (reason: string, description: any) => {
+      log.debug("got socket close reason=%s description=%s messages=%s",
+        reason, description?.message ?? description, [...this._messageCallbacks.keys()]);
+
+      const err = description ?? new Error(reason);
+      for(const msgNum of this._messageCallbacks.keys()) {
+        const cb = this._messageCallbacks.get(msgNum);
+        if (cb) {
+          log.debug("calling cb for msg %d with %s", msgNum, err?.message || err);
+          cb(err);
+        }
+      }
+      this._messageCallbacks.clear();
+
+      log.debug("calling close handler");
+      handler();
+    };
+    this._socket.on('close', wrappedHandler);
+    this._eventHandlers.push({event: 'close', handler: wrappedHandler});
+  }
+
+  public set onmessage(handler: (msg: string) => void) {
+    const wrappedHandler = (msg: Buffer) => {
+      log.debug("got message (%d bytes) %s", msg.length, msg.slice(0, 100));
+      handler(msg.toString());
+    };
+    this._socket.on('message', wrappedHandler);
+    this._eventHandlers.push({event: 'message', handler: wrappedHandler});
+  }
+
+  public removeAllListeners() {
+    for (const {event, handler} of this._eventHandlers) {
+      this._socket.off(event, handler);
+    }
+    this._eventHandlers = [];
+  }
+
+  public close() {
+    this._socket.close();
+  }
+
+  public terminate() {
+    this._socket.close(/*discard*/ true);
+  }
+
+  public get isOpen() {
+    return this._socket.readyState === 'open';
+  }
+
+  public send(data: string, cb: (err?: Error) => void) {
+    const msgNum = this._messageCounter++;
+    this._messageCallbacks.set(msgNum, cb);
+    log.debug("sending msg %d: %s", msgNum, { toString:()=>data.slice(0, 100) });
+    this._socket.send(data, {}, () => {
+      this._messageCallbacks.delete(msgNum);
+      log.debug("calling cb for msg %d", msgNum);
+      cb();
+    });
+  }
+}
+
+export class GristServerSocketWS extends GristServerSocket {
+  private _eventHandlers: Array<{event: string, handler: (...args: any[]) => void}> = [];
+  private _messageCounter = 0;
+
+  constructor(private _ws: WS) { super(); }
+
+  public set onerror(handler: (err: unknown) => void) {
+    this._ws.on('error', handler);
+    this._eventHandlers.push({event: 'error', handler});
+  }
+
+  public set onclose(handler: () => void) {
+    this._ws.on('close', handler);
+    this._eventHandlers.push({event: 'close', handler});
+  }
+
+  public set onmessage(handler: (msg: string) => void) {
+    this._ws.on('message', (msg: Buffer) => handler(msg.toString()));
```

Comment:
Looking into it, `ws` used to emit a `string` for text messages but this was [changed in 8.0.0](https://github.com/websockets/ws/pull/1904). It is now documented to [only emit Buffers](https://github.com/websockets/ws/blame/master/doc/ws.md#event-message). This went undetected when you upgraded because `JSON.parse` silently calls `toString()`.

Ah, I see that when you upgraded `ws` in https://github.com/gristlabs/grist-core/pull/614, `@types/ws` stayed at `^6` so the types do not reflect the change. I have added the upgrade to `^8` to this PR.

I also went ahead and made `onmessage` in `GristClientSocket` always emit a `string` instead of unnecessarily emulating an event with a `data` property.